File: app/src/test/java/org/metabrainz/mobile/LookupViewModelTest.java
package org.metabrainz.mobile;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;

import org.junit.Rule;
import org.junit.Test;
import org.metabrainz.mobile.data.sources.api.entities.WikiSummary;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Artist;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Label;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Recording;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Release;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.ReleaseGroup;
import org.metabrainz.mobile.presentation.features.artist.ArtistViewModel;
import org.metabrainz.mobile.presentation.features.label.LabelViewModel;
import org.metabrainz.mobile.presentation.features.recording.RecordingViewModel;
import org.metabrainz.mobile.presentation.features.release.ReleaseViewModel;
import org.metabrainz.mobile.presentation.features.release_group.ReleaseGroupViewModel;
import org.metabrainz.mobile.util.Resource;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.metabrainz.mobile.AssertionUtils.checkArtistAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkLabelAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkRecordingAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkReleaseAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkReleaseGroupAssertions;
import static org.metabrainz.mobile.AssertionUtils.checkWikiAssertions;
import static org.metabrainz.mobile.EntityTestUtils.getTestArtist;
import static org.metabrainz.mobile.EntityTestUtils.getTestArtistMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestArtistWiki;
import static org.metabrainz.mobile.EntityTestUtils.getTestLabel;
import static org.metabrainz.mobile.EntityTestUtils.getTestLabelMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestRecording;
import static org.metabrainz.mobile.EntityTestUtils.getTestRecordingMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestRelease;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseGroup;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseGroupMBID;
import static org.metabrainz.mobile.EntityTestUtils.getTestReleaseMBID;
import static org.metabrainz.mobile.LiveDataTestUtil.getOrAwaitValue;

public class LookupViewModelTest {

    @Rule
    public InstantTaskExecutorRule executorRule = new InstantTaskExecutorRule();

    @Test
    public void testArtistViewModel() {
        Artist testArtist = getTestArtist();
        WikiSummary testWiki = getTestArtistWiki();
        ArtistViewModel viewModel = new ArtistViewModel(new MockLookupRepository());
        viewModel.setMBID(getTestArtistMBID());
        try {
            Resource<Artist> resource = getOrAwaitValue(viewModel.getData());
            assertEquals(Resource.Status.SUCCESS, resource.getStatus());
            checkArtistAssertions(testArtist, resource.getData());

            Resource<WikiSummary> wikiResource = getOrAwaitValue(viewModel.getWikiData());
            assertEquals(Resource.Status.SUCCESS, wikiResource.getStatus());
            checkWikiAssertions(testWiki, wikiResource.getData());
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testReleaseViewModel() {
        Release testRelease = getTestRelease();
        ReleaseViewModel viewModel = new ReleaseViewModel(new MockLookupRepository());
        viewModel.setMBID(getTestReleaseMBID());
        try {
            Resource<Release> resource = getOrAwaitValue(viewModel.getData());
            assertEquals(Resource.Status.SUCCESS, resource.getStatus());
            checkReleaseAssertions(testRelease, resource.getData());
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testReleaseGroupViewModel() {
        ReleaseGroup testReleaseGroup = getTestReleaseGroup();
        ReleaseGroupViewModel viewModel = new ReleaseGroupViewModel(new MockLookupRepository());
        viewModel.setMBID(getTestReleaseGroupMBID());
        try {
            Resource<ReleaseGroup> resource = getOrAwaitValue(viewModel.getData());
            assertEquals(Resource.Status.SUCCESS, resource.getStatus());
            checkReleaseGroupAssertions(testReleaseGroup, resource.getData());
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testLabelViewModel() {
        Label testLabel = getTestLabel();
        LabelViewModel viewModel = new LabelViewModel(new MockLookupRepository());
        viewModel.setMBID(getTestLabelMBID());
        try {
            Resource<Label> resource = getOrAwaitValue(viewModel.getData());
            assertEquals(Resource.Status.SUCCESS, resource.getStatus());
            checkLabelAssertions(testLabel, resource.getData());
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testRecordingViewModel() {
        Recording testRecording = getTestRecording();
        RecordingViewModel viewModel = new RecordingViewModel(new MockLookupRepository());
        viewModel.setMBID(getTestRecordingMBID());
        try {
            Resource<Recording> resource = getOrAwaitValue(viewModel.getData());
            assertEquals(Resource.Status.SUCCESS, resource.getStatus());
            checkRecordingAssertions(testRecording, resource.getData());
        } catch (Exception e) {
            e.printStackTrace();
            fail();
        }
    }

}

File: app/src/test/java/org/metabrainz/mobile/LookupViewModelTest.kt
package org.metabrainz.mobile

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.TestCoroutineDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runBlockingTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.metabrainz.mobile.AssertionUtils.checkArtistAssertions
import org.metabrainz.mobile.AssertionUtils.checkLabelAssertions
import org.metabrainz.mobile.AssertionUtils.checkRecordingAssertions
import org.metabrainz.mobile.AssertionUtils.checkReleaseAssertions
import org.metabrainz.mobile.AssertionUtils.checkReleaseGroupAssertions
import org.metabrainz.mobile.AssertionUtils.checkWikiAssertions
import org.metabrainz.mobile.EntityTestUtils.testArtist
import org.metabrainz.mobile.LiveDataTestUtil.getOrAwaitValue
import org.metabrainz.mobile.presentation.features.artist.ArtistViewModel
import org.metabrainz.mobile.presentation.features.label.LabelViewModel
import org.metabrainz.mobile.presentation.features.recording.RecordingViewModel
import org.metabrainz.mobile.presentation.features.release.ReleaseViewModel
import org.metabrainz.mobile.presentation.features.release_group.ReleaseGroupViewModel
import org.metabrainz.mobile.util.Resource.Status.SUCCESS
import org.metabrainz.mobile.EntityTestUtils.testArtistWiki as testArtistWiki

class LookupViewModelTest {

    // FIXME: flaky tests because we use Dispatchers.IO in running the coroutines in production
    // but the test library only provides option to replace Dispatchers.Main. Solution can be to
    // inject the CoroutineScope in the ViewModel.
    @ExperimentalCoroutinesApi
    val testDispatcher = TestCoroutineDispatcher()

    @get:Rule
    val executorRule = InstantTaskExecutorRule()

    @ExperimentalCoroutinesApi
    @Before
    fun setup() {
        Dispatchers.setMain(testDispatcher)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testArtistViewModel() = testDispatcher.runBlockingTest {
        val testArtist = testArtist
        val testWiki = testArtistWiki
        val viewModel = ArtistViewModel(MockLookupRepository())
        viewModel.mbid.value = EntityTestUtils.testArtistMBID
        val resource = getOrAwaitValue(viewModel.data)
        assertEquals(SUCCESS, resource?.status)
        checkArtistAssertions(testArtist, resource?.data!!)
        val wikiResource = getOrAwaitValue(viewModel.wikiData)
        assertEquals(SUCCESS, wikiResource?.status)
        checkWikiAssertions(testWiki, wikiResource?.data!!)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testReleaseViewModel() = testDispatcher.runBlockingTest {
        val testRelease = EntityTestUtils.testRelease
        val viewModel = ReleaseViewModel(MockLookupRepository())
        viewModel.mbid.value = EntityTestUtils.testReleaseMBID
        val resource = getOrAwaitValue(viewModel.data)
        assertEquals(SUCCESS, resource?.status)
        checkReleaseAssertions(testRelease, resource?.data!!)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testReleaseGroupViewModel() = testDispatcher.runBlockingTest {
        val testReleaseGroup = EntityTestUtils.testReleaseGroup
        val viewModel = ReleaseGroupViewModel(MockLookupRepository())
        viewModel.mbid.value = EntityTestUtils.testReleaseGroupMBID
        val resource = getOrAwaitValue(viewModel.data)
        assertEquals(SUCCESS, resource?.status!!)
        checkReleaseGroupAssertions(testReleaseGroup, resource.data!!)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testLabelViewModel() = testDispatcher.runBlockingTest {
        val testLabel = EntityTestUtils.testLabel
        val viewModel = LabelViewModel(MockLookupRepository())
        viewModel.mbid.value = EntityTestUtils.testLabelMBID
        val resource = getOrAwaitValue(viewModel.data)
        assertEquals(SUCCESS, resource?.status)
        checkLabelAssertions(testLabel, resource?.data!!)
    }

    @ExperimentalCoroutinesApi
    @Test
    fun testRecordingViewModel() = testDispatcher.runBlockingTest {
        val testRecording = EntityTestUtils.testRecording
        val viewModel = RecordingViewModel(MockLookupRepository())
        viewModel.mbid.value = EntityTestUtils.testRecordingMBID
        val resource = getOrAwaitValue(viewModel.data)
        assertEquals(SUCCESS, resource?.status!!)
        checkRecordingAssertions(testRecording, resource.data!!)
    }

    @ExperimentalCoroutinesApi
    @After
    fun teardown() {
        Dispatchers.resetMain()
        testDispatcher.cleanupTestCoroutines()
    }
}